NIFI-5984: Enabled Kerberos Authentication for PutKudu

markap14 opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

jtstorck commented:
Reviewing...

jtstorck commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/pom.xml:
```diff
@@ -59,20 +58,24 @@
+            <version>18.0</version>
         </dependency>
+
+        <dependency>
```
This dependency can be removed, since Kudu doesn't use any of the "Hadoop Resources" like the HDFS/Hive/HBase components, and you're only referencing UGI from hadoop-commons. The nifi-security-utils module has classes in the org.apache.nifi.security.krb package that can be used to log in with a KDC, perform PrivelegedExceptionActions via the KerberosAction class, and handle automatic relogins when tickets expire.  I have a PR up now that adds some usability modifications to KerberosAction.  You could base this PR off of my Kerberized DBCPConnectionPool branch (https://github.com/jtstorck/nifi/tree/NIFI-5985) to work with those changes.

markap14 replied:
That looks reasonable. I updated the JIRA to indicate that it is blocked by NIFI-5985.

jtstorck commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/PutKudu.java:
```diff
@@ -181,21 +199,37 @@
+        final String keytab = credentialsService.getKeytab();
+        final String principal = credentialsService.getPrincipal();
+
+        // We do not have any Hadoop Configuration to use for this Processor. However, if we call UserGroupInformation.loginUserFromKeytab() it will return
```
Based on my previous comment about using KerberosAction in nifi-security-utils package org.apache.nifi.security.krb, along with KerberosUser, you can remove the dependency on hadoop-common.  Please see my Kerberized DBCPConnectionPool branch (https://github.com/jtstorck/nifi/tree/NIFI-5985) for an example.

jtstorck replied:
Since I can't leave a comment on a line that was not modified in the PR, I'll add it here.  I think it's also important when the processor is stopped to log the principal out of the KDC, in the closeClient method.

jtstorck commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/test/java/org/apache/nifi/processors/kudu/MockPutKudu.java:
```diff
@@ -61,7 +62,7 @@ protected Upsert upsertRecordToKudu(KuduTable kuduTable, Record record, List<Str
 
     @Override
-    protected KuduClient createClient(final String masters) {
+    protected KuduClient createClient(final String masters, final KerberosCredentialsService credentialsService) {
```
Could you add a test that covers the logic to create secure/unsecure KuduClients?

jtstorck commented:
> The code LGTM, I'm not able to give it a try right now unfortunately. Just one question though: does it take care of the ticket renewal? or is the underlying implementation taking care of it? (asking since the client is initialized in the OnScheduled method)

UGI will take care of the relogin, but I think only if the UGI.doAs is used to execute the action.  If the PR is updated to use KerberosAction from nifi-security-utils org.apache.nifi.security.krb.KerberosAction, it will handle expired tickets.

markap14 commented:
@jtstorck I pushed another commit that I believe addresses all review feedback and rebases against master in order to resolve a conflict that was introduced.

jtstorck commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/PutKudu.java:
```diff
@@ -206,21 +227,48 @@ public void OnScheduled(final ProcessContext context) throws KuduException {
+        final String principal = credentialsService.getPrincipal();
+        kerberosUser = loginKerberosUser(principal, keytab);
+
+        final KerberosAction<KuduClient> kerberosAction = new KerberosAction<>(kerberosUser, () -> buildClient(masters), getLogger());
```
In addition to creating the client with KerberosAction, I would recommend running the invocations of methods on the created Kudu client with KerberosAction as well, to make sure the correct Subject is available.  We did the same with the HDFS, Solr, Hive, HBase, etc, processors.

Specific to KerberosAction, you can see an example of that in `SolrProcessor`'s `onTrigger` method.

markap14 replied:
Thanks, will do.

jtstorck commented:
Thanks for the updates and this contribution, @markap14!

Tested this against a kerberized cluster (HDFS, Impala, Kudu), creating Impala tables stored in Kudu, inserting data into Kudu with the PutKudu processor configured to use a KerberosCredentialsService.

LGTM, +1.  Merging to master!

zorel commented:
Hello.

 I still have issues with expired tickets, I have to restart the processor every week with nifi 1.9.2

granthenke commented:
@zorel Is this still an issue for you? If so we should file a jira.

zorel commented:
> @zorel Is this still an issue for you? If so we should file a jira.

Yeah, I'm not part of the team anymore, but it seems they still have the problem, every week. I'll see with them if/when they upgrade to 1.10 to see if it still happens. Thanks
